presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java

```diff
@@ -0,0 +1,502 @@
+            }
+            if (blocked != null && !blocked.isDone()) {
+                return blocked;
+            }
+            blocked = Futures.allAsList(mergeSources.stream().map(MergeSource::isBlocked).collect(Collectors.toList()));
+            return blocked;
+        }
+
+        public boolean isFinished()
+        {
+            return mergeSources.stream().allMatch(MergeSource::isFinished);
+        }
+
+        public List<MergeSource> getMergeSources()
+        {
+            return mergeSources;
+        }
+    }
+
+    private static class MergeSource
+    {
+        private final URI location;
+        private final ExchangeClient exchangeClient;
+        private final PagesSerde serde;
+
+        private PageWithPosition currentPage;
+        private ListenableFuture<?> blocked;
+
+        public MergeSource(URI location, ExchangeClient exchangeClient, PagesSerde serde)
+        {
+            this.location = requireNonNull(location, "location is null");
+            this.exchangeClient = requireNonNull(exchangeClient, "exchangeClient is null");
+            this.serde = requireNonNull(serde, "serde is null");
+        }
+
+        public URI getLocation()
+        {
+            return location;
+        }
+
+        public ListenableFuture<?> isBlocked()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return NOT_BLOCKED;
+            }
+            if (blocked != null && !blocked.isDone()) {
+                return blocked;
+            }
+            blocked = exchangeClient.isBlocked();
+            return blocked;
+        }
+
+        public boolean isFinished()
+        {
+            return exchangeClient.isFinished() && (currentPage == null || currentPage.isFinished());
+        }
+
+        @Nullable
+        public PageWithPosition getPage()
+        {
+            if (currentPage != null && !currentPage.isFinished()) {
+                return currentPage;
+            }
+
+            checkState(blocked != null, "blocked is null");
+            if (!isBlocked().isDone()) {
+                return null;
+            }
+
+            if (exchangeClient.isFinished()) {
+                return null;
+            }
+
+            SerializedPage serializedPage = exchangeClient.pollPage();
+            checkState(serializedPage != null, "exchange client has returned null for the next page");
+            Page page = serde.deserialize(serializedPage);
+            currentPage = new PageWithPosition(page);
+            return currentPage;
+        }
+    }
+
+    private static class PageWithPosition
+    {
+        private final Page page;
+        private int position = 0;
+
+        public PageWithPosition(Page page)
+        {
+            this.page = requireNonNull(page, "page is null");
+        }
+
+        public Page getPage()
+        {
+            return page;
+        }
+
+        public int getPosition()
+        {
+            checkPosition();
```

Comment:
Not needed here.
`checkPosition` should only be placed in places where the position or `Page` may change (`incrementPosition` and c-tor).